controllers/machine_controller_node_labels.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	apicorev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/controllers/external"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *MachineReconciler) reconcileInterruptibleNodeLabel(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) (ctrl.Result, error) {
+	logger := ctrl.LoggerFrom(ctx)
+
+	// Check that the Machine hasn't been deleted or in the process.
+	if !machine.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	// Check that the Machine has a NodeRef.
+	if machine.Status.NodeRef == nil {
+		return ctrl.Result{}, nil
+	}
+
+	// Get the infrastructure object
+	infra, err := external.Get(ctx, r.Client, &machine.Spec.InfrastructureRef, machine.Namespace)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Get interruptible instance status from the infrastructure provider.
+	interruptible, _, err := unstructured.NestedBool(infra.Object, "status", "interruptible")
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get interruptible status from infrastructure provider for Machine %q in namespace %q: %w", machine.Name, machine.Namespace, err)
+	}
+
+	if !interruptible {
+		return ctrl.Result{}, nil
+	}
+
+	remoteClient, err := r.Tracker.GetClient(ctx, util.ObjectKey(cluster))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.setInterruptibleNodeLabel(ctx, remoteClient, machine.Status.NodeRef.Name)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Set interruptible label to Machine's Node", "nodename", machine.Status.NodeRef.Name)
+	r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetInterruptibleNodeLabel", machine.Status.NodeRef.Name)
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MachineReconciler) setInterruptibleNodeLabel(ctx context.Context, remoteClient client.Client, nodeName string) error {
+	node := &apicorev1.Node{}
+
+	err := remoteClient.Get(ctx, client.ObjectKey{Name: nodeName}, node)
+	if err != nil {
+		return err
+	}
+
```

Comment:
We should check here if the label is already set, and return early if so.

Reply:
I think this is covered by L88-90?